File: perl/fact.pl
use 5.16.0;
sub fact {
    my $acc=sub{my($n,$fact)=@_;
		if ($n){
		    __SUB__->($n-1,$fact*$n);}
		else {$fact;}
    };
    $acc->(@_,1);
}
my $ans=fact(@ARGV);
say $ans;

File: rust/old/snippits/fact.rs
#[no_mangle]
pub fn fact(n: u64) -> u64 {
    if n <= 1 {
        1
    } else {
        n*fact(n-1)
    }
}
#[no_mangle]
pub fn fact_tail(n: u64) -> u64 {
    fn f(n: u64, acc:u64) -> u64 { if n<=1 {acc} else {f(n-1,n*acc)}}
    f(n, 1)
}
